server: option to wait for default_target_cluster on connect

stevendanna opened the pull request:
Virtual cluster startup is asynchronous. As a result, it can be hard to
orchestrate bringing a virtual cluster up since the client needs to
retry until the cluster is available.

This makes it a little easier by optionally allowing connections to
the default_target_cluster to be held until the cluster is online.

As a result, a user who does:

    ALTER VIRTUAL CLUSTER app START SERVICE SHARED
    SET CLUSTER SETTING server.controller.default_target_cluster = 'app'

Should be able to immediately connect to the same node without having
to handle retries while the virtual cluster starts up.

Note that we can extend this to wait for _any_
tenant. The restriction on only waiting on the default_target_cluster
is that we need to thread a little state into the server controller in
order to avoid pitfalls such as holding a lot of connections open for
longer than needed for a tenant that will never become available
because it doesn't exist.

In a future PR, we will set this option by default in the replication-source and replication-target
configuration profiles.

Built on #113666

Informs https://github.com/cockroachdb/cockroach/issues/111637

Release note: None

stevendanna committed:
server: wait for the default_target_cluster on connect

Virtual cluster startup is asynchronous. As a result, it can be hard to
orchestrate bringing a virtual cluster up since the client needs to
retry until the cluster is available.

This makes it a little easier by optionally allowing connections to
the default_target_cluster to be held until the cluster is online.

As a result, a user who does:

    ALTER VIRTUAL CLUSTER app START SERVICE SHARED
    SET CLUSTER SETTING server.controller.default_target_cluster = 'app'

Should be able to immediately connect to the same node without having
to handle retries while the virtual cluster starts up.

Note that in future PRs we can extend this to wait for _any_
tenant. The restriction on only waiting on the default_target_cluster
is that we need to thread a little state into the server controller in
order to avoid pitfalls such as holding a lot of connections open for
longer than needed for a tenant that will never become available
because it doesn't exist.

Informs #111637

Release note: None

stevendanna committed:
server: use singleflight when waiting on sql server

Using singleflight avoids having a large number of connections polling
for the tenant to start up.

Release note: none

stevendanna committed:
server: avoid polling when waiting for tenants

Rather than polling, this uses a channel to notify us when the set of
tenants has changed or when the tenant's state has changed.

Release note: None

stevendanna committed:
server: use RWMutex in server ochestrator

We call getServer on every new connection. It seem prudent that we
wouldn't want concurrent connections to contend over a mutex when most
of the time nothing is writing to the data protected by this mutex.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/113687)
<!-- Reviewable:end -->

stevendanna commented:
I recommend going commit-by-commit during the review since none of the steps are too complicated even though the overall diff ended up a bit large.

stevendanna commented:
Pretty sure that the test failed because we are slow enough during stress that the 10s timeout isn't enough. I'll dig into this but I don't expect it to substantially change the shape of this.

yuzefovich reviewed (commented):
Seems good to me.

> Reviewed 3 of 3 files at r1, 3 of 3 files at r2, 3 of 3 files at r3, 8 of 8 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-:-NiGQ9CX28RDRy86h6Jx:b-femq48)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @msbutler, and @stevendanna)

___
*[`pkg/multitenant/tenant_config.go` line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-NiGN2cA6gIijlE3sK8a:-NiGN2cA6gIijlE3sK8b:botjjcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b075a81254959e19e9b0f6c2108f665af724db3/pkg/multitenant/tenant_config.go#L46)):*
> ```Go
> )
> 
> // WaitForClusterStart, if enabled, instructs the tenant controller to
> ```

What's the benefit of having two settings rather than a single duration setting in which 0 means "disabled"? Also, why not enable it by default?
___
*[`pkg/multitenant/tenant_config.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-NiGNAv51TwZ9w4dqkPd:-NiGNAv51TwZ9w4dqkPe:b9nyyvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b075a81254959e19e9b0f6c2108f665af724db3/pkg/multitenant/tenant_config.go#L47)):*
> ```Go
> 
> // WaitForClusterStart, if enabled, instructs the tenant controller to
> // wait up to WaitForClusterStartTimeout for the defuault virtual
> ```

nit: `s/defuault/default/`.
___
*[`pkg/multitenant/tenant_config.go` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-NiGNDzQCb9qE8ZFA38A:-NiGNDzQCb9qE8ZFA38B:bki1qmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b075a81254959e19e9b0f6c2108f665af724db3/pkg/multitenant/tenant_config.go#L56)):*
> ```Go
> )
> 
> // WaitForClusterStartTimeout is the amoutn of time the the tenant
> ```

nit: `s/amoutn/amount/` and `s/the the/the/`.
___
*[`pkg/server/server_controller_accessors.go` line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-NiGOz11CeHn4gsVvqmP:-NiGOz11CeHn4gsVvqmQ:b-4bx0gs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8ed881050be8443a6f6c759dbbb66c7e2b57bfc/pkg/server/server_controller_accessors.go#L21)):*
> ```Go
> 
> // getServer retrieves a reference to the current server for the given
> // tenant name.
> ```

nit: mention what the returned channel is about.


<!-- Sent from Reviewable.io -->

msbutler commented on pkg/server/server_controller_sql.go:
we should only wait if the err matches `server for tenant %q not ready`, correct? If so,  I feel like this error string should have its own mark.

stevendanna replied:
No, we want to retry on both errors. Both the existence of the tenant and the startup is async from the perspective of the server controller because we learn about tenants via a rangefeed.

msbutler commented on pkg/server/server_controller_sql.go:
naive question: did you consider using `tenantWaiter.DoChan()` to orchestrate this dance instead? At a high level, its purpose seems to match the use case here, but there may be details I'm missing.

msbutler reviewed (commented):
A fun exploration of server code!

stevendanna commented on pkg/server/server_controller_test.go:
```diff
@@ -124,6 +124,37 @@ func TestServerControllerStopStart(t *testing.T) {
+	sqlRunner.Exec(t, "CREATE TENANT hello")
+	sqlRunner.Exec(t, "ALTER VIRTUAL CLUSTER hello START SERVICE SHARED")
+	sqlRunner.Exec(t, "SET CLUSTER SETTING server.controller.default_target_cluster = 'hello'")
+	require.NoError(t, tryConnect())
```
Yes, but only rarely.

stevendanna committed:
server: use a single setting for tenant waiting

We can use a zero-wait time to indicate no waiting, saving a cluster
setting.

Epic: none

Release note: None

stevendanna committed:
server: use singleflight.DoChan to avoid shared cancellation

Release note: None

stevendanna reviewed (commented):
Thanks for the reviews. Should be ready for another look. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-:-NiZle9BDlBPXri8lLOz:bupnq3r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @msbutler, and @yuzefovich)

___
*[`pkg/multitenant/tenant_config.go` line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-NiGN2cA6gIijlE3sK8a:-NiZlLlu3WjyYUZXeGW7:bmrzgzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b075a81254959e19e9b0f6c2108f665af724db3/pkg/multitenant/tenant_config.go#L46)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

What's the benefit of having two settings rather than a single duration setting in which 0 means "disabled"? Also, why not enable it by default?
</blockquote></details>

I agree, one option would be better. Done.  I've changed it to enabled by default, but depending on how long it takes me to get this in perhaps we backport it disabled.
___
*[`pkg/multitenant/tenant_config.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-NiGNAv51TwZ9w4dqkPd:-NiZlRde4pJKk2q5yCZ2:b-i21bbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b075a81254959e19e9b0f6c2108f665af724db3/pkg/multitenant/tenant_config.go#L47)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/defuault/default/`.
</blockquote></details>

Thanks. Done.
___
*[`pkg/multitenant/tenant_config.go` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-NiGNDzQCb9qE8ZFA38A:-NiZlSbD8SRzvHOh-xdX:b-i21bbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b075a81254959e19e9b0f6c2108f665af724db3/pkg/multitenant/tenant_config.go#L56)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/amoutn/amount/` and `s/the the/the/`.
</blockquote></details>

Thanks. Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 14 of 14 files at r6, 3 of 3 files at r7, 3 of 3 files at r8, 8 of 8 files at r9, 3 of 3 files at r10, 3 of 3 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-:-Ni_HMg99ml1E347jlJW:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @msbutler, and @stevendanna)

___
*[`pkg/multitenant/tenant_config.go` line 48 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113687#-Ni_DD-7AKYW8OkQj6W1:-Ni_DD-7AKYW8OkQj6W2:b12krp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/95c4679f124fac8cde7f0aad776573568a484410/pkg/multitenant/tenant_config.go#L48)):*
> ```Go
> // WaitForClusterStartTimeout is the amount of time the tenant
> // controller will wait for the default virtual cluster to have an
> // active SQL server, if WaitForClusterStart is true.
> ```

nit: `WaitForClusterStart` is no more.


<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r=yuzefovich
